### DISCUSSION TITLE
Creating Your Own Fakes Instead of Using Mocking Libraries

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#82] [General]
# Creating Your Own Fakes Instead of Using Mocking Libraries

With respect to the current topic around Moq, I want to showcase how you can easily roll out your own fakes so that you are not depending on a third party library.¨

https://steven-giesel.com/blogPost/2ad414c8-273f-4976-a9de-c95f0e791d51

**Comment by vertonghenb** [comment]
Nice article about Faking data, I've dedicated some slides to it in our C# course, "Fake it 'till you make it" using bogus. 
https://hogent-web.github.io/csharp/chapters/06/slides/index.html#73

**Comment by toupswork** [comment]
There's something I've been wanting to try. Rather than implementing them with a class, I want to try creating an interface that inherits from the real interface and implements the methods now that interfaces allow this, then pass that into the class you're testing

**Reply by linkdotnet** [reply]
What would be the advantage in the grand scheme?
Where would that deriving interface live?

**Reply by toupswork** [reply]
In the unit test project. I actually tried it today. It didn't work out the way I hoped. It still requires that you implement a class; I thought maybe I could pass in a null reference, an object force-casted to the interface by using Unsafe.As<T>, or a Dispatch proxy. But the class still tries to call the method on the original interface and not the child interface, even if you specify "new" on the methods. It was worth a try, but I'm still better off with Moq.  What I am really looking forward to is when someone releases a tool that allows you to use interceptors for mocking.. That will be great.. I think writing unit tests is the worse part of programming, next to trying to understand someone else's terrible code.

**Reply by linkdotnet** [reply]
That is a nice idea.
Currently interceptors are nice but they need polishing and better APIs specifically designed for such use cases.

I created on for one of the libraries where I am maintainer, but as a dev you are left alone.
There is almost no documentation besides one cookbook from Microsoft. No XMLDOC for the many of the new API.
I had to use much trial and error to get it working (hopefully stable).

I hope .NET 9 will continue the journey and move then towards stability (and way better docs).

**Reply by toupswork** [reply]
Yeah. It feels like a quick hack they through in to get Native AOT out the door. Microsoft has been hemorrhaging customers because of the cold-start time problem with .NET powered lambdas and Azure functions. People are dumping .NET for Go, Rust, and Python. Where I work, we had some teams try Lambdas written in .NET and they were taking way too long to start. I wish I could use it so bad but we use DataDog and they don't support AOT yet... So angry about that.

**Reply by linkdotnet** [reply]
Python is faster in startup than .NET?
I always heard it the other was around. People ditching Python in favour of other languages closer tot he underlying hardware to have a smaller bill at the end of the day.

with the current plan of MS to be more suited for cloud native feature I hope they invest some more resources in the whole topic!

**Reply by toupswork** [reply]
I only ever wrote lambdas using Python. And I never had a cold-start issue.  I am not surprised that .NET is slower than Python when it comes to short-lived apps only.  Traditionally, they focused on long-lived application performance. They didn't historically care too much about how long it took to start up, because not only does the JIT have impressive optimizations when it compiles the IL to native, but it also observes the behavior of the application over time and then based on that it will re-compile it again with more optimizations that are customized based on usage patterns, so that, in some cases, the Jitted code can be faster than native code. Then they went to town with reflection, because again, the reflection would run at startup, but it would get cached so you would only get that first hit and it would be fast from then on.  Reflection is mostly all what MVC is, because you get that nice declarative programming with attributes and things just wired themselves up like magic and developers just ate it up. Then all the Nuget package utilities started doing the same thing, where you can just pass in the Executing assembly and it will wire up everything. Now they're paying for that, and that's why only minimal APIs are supported with AOT.  Also, from the early days, the GC was configured in a server environment to just let the workingset memory eat up as much memory as it wants and put off GC collection, so you create a simple hello world API and it takes up 100MB. So yeah, that's how we got here. But now .NET 8 is introducing alternatives to reflection and it allows you to configure GC pressure.. They're so behind now. They had Project N a long time ago where they were experimenting with AOT native compilation, and they're still late to the game.

**Reply by linkdotnet** [reply]
> Also, from the early days, the GC was configured in a server environment to just let the workingset memory eat up as much memory

isn’t that still somewhat true?
While they added a bit stuff to control the GC more fine-grained, the overarching architecture is the same.

Sometimes they add bits like the Frozen Heap where the new „frozen“ stuff hangs around, but these are tweaks of special cases rather than overarching things.

„Unfortunately“ I am working less in the cloud these days where this becomes an issue.
Still I follow closely what they are doing.

Depending on the use case I would argue golang is also often a good choice for the cloud environment.
I do like Googles approach here.

**Reply by toupswork** [reply]
Yeah, but you can now set an upper limit and control how aggressive it is at collecting. I haven't read up on the details of the configuration options, but I know that you can tell it not to grow beyond a certain size and that it allows you to make it behave the way it does on a workstation. I was helping out another team a few months back and they were doing AWS stuff.. I even learned PostgreSQL and I think it's awesome, after being a MS fan boy my whole life and using mostly SQL.  Now I am back on my team and we're building a React App with a .NET 8 API running in on-prem kubernetes. I liked working in the cloud and hope we get a project later on so I can go back to it because I learned so much.  I was going to mention GoLang. It looks neat. I know that a lot of .NET shops moved to it because of the cold-start issue.  It's a compiled language, which is cool, and it's one higher abstraction order above Rust. But when it comes to a general purpose language, .NET in my opinion is the best. I'm also a huge TypeScript fan. Hard to say which language I like more: TS or C#.

**Reply by linkdotnet** [reply]
I love the structural typing of TypeScript - and to that extent of Golang as well.
I wish I could use that (more often) in .NET.

A lot of things are somewhat duck typed in C# like foreach, await and using where you don’t necessarily need the interface - it just makes it painfully obvious what your intend is.

Postgresql is a beast. In the projects I work you mainly see MSSQL Server. While rock solid you can see how it tried to onboard features that Postgres has since ages.

**Reply by toupswork** [reply]
Agreed. There's something about TypeScript that makes me love it. Maybe cause the advanced stuff looks so cryptic, and I wanted a challenge. I don't know, but I love building components with TypeScript where as you start adding properties, the type inference flows out and you get all this cool auto completion support, and I can make it really hard to pass the wrong data into it. I hate the `any` keyword, though. So many people don't try to learn TS and as soon as they get stuck, they put in `any`.

**Reply by toupswork** [reply]
Yes, agreed on the duck typing. They need to stop that.  Do you remember Global.aspx duck typing? yuck.  And yea, PostgreSQL is a beast. It's so freaking powerful and eloquent.

**Reply by linkdotnet** [reply]
Oh the good old Global.asax
Yes it was weird. Weird conventions that didn’t really fit into .net at all.

I do agree - using any everywhere gives you the overhead of typescript but your still using JavaScript.
Sometimes big 3rd component libraries (in the angular space) don’t really type their API so it is really hard that it doesn’t bleed into your own code. 
And sometimes even worse: they use the wrong type:)
